### ISSUE TITLE
1986ВЕ1Т и ethernet

### ISSUE THREAD

**Issue by aol-nnov** [sergev/uos-embedded#18]
@sergev подскажите, пожалуйста по поводу сабжа?
на стейбле есть два драйвера: "старый" и "новый" - какова их замысловатая судьба? Дело в том, что ни с одним из них на демо-борде не получается..

Так же, я обнаружил ветку `stable-eth-devel`, но там вообще странные структуры описываются, я таких в документации к процессору даже не вижу (например, [восьмибайтовый хэш](https://github.com/sergev/uos-embedded/blob/stable-eth-devel/sources/runtime/cortex-m/io-1986ve1.h#L1976))

У меня на борде присутствует дополнительный социллятор на 25 МГц - видимо, это уже "после второй ревизии" (чего бы это ни значило - погружаюсь в эту тему вторую неделю лишь, ооочень много противоречивой информации)

Справедливости ради, хочу отметить, что заливал "кейловский пример", который на компакт диске есть и в hex, я залил его через stlink - он работает, но он на кейле, которого у меня нет. 

Сейчас пытаюсь провести параллели между константами в cmsis и uos-ом, пока туго идет.

Понавтыкал отладочной печати в разные места `net`, складывается впечатление, что до `ip`доходят только широковещательные пакеты, причем, как со "старым" драйвером, так и с "новым".

Заранее благодарен за любые подсказки!

**Comment by aol-nnov** [comment]
попутно через отладчик выяснил следующее:

на `stable` пример `examples/arm-1986be1/test-ve1eth.c` виснет на [вот этой строке](https://github.com/sergev/uos-embedded/blob/stable/sources/milandr/ve1eth.c#L1344)

однако, если `sources/kernel` откатить на ветку `stable-eth-dev`, то залочка на этом мьютексе работает, как надо и все таски получают время на свою работу.

Из диффа `sources/kernel` видно, что на `stable` внутрь мьютексов [добавлены](https://github.com/sergev/uos-embedded/compare/stable-eth-devel...stable#diff-ce977251d147401c687a186f68b552ecR104) поля `active` и `saved_msg` и дополнительный код для их обработки.

Немного непонятно, в каком направлении двигаться в свете найденного..

**Comment by aol-nnov** [comment]
позволил себе небольшую шалось в примере:
```
uint8_t my_macaddr[] = {0xff,0xff,0xff,0xff,0xff,0xff};
eth_init(eth, "eth0", 80, &pool, arp, my_macaddr, ARM_ETH_PHY_FULL_AUTO);
```
и, о чудо (с) (тм), пинги пошли и смог подключиться к tcp серверу телнетом.

когда ставишь более корректный мак, работает, как я описывал выше.

**Comment by vatilin** [comment]
Андрей, добрый день!

По поводу изменений в ядре - это моя ошибка. Экспериментальный код,
нужный только в одном проекте, попал в репозиторий. Прошу прощения!
Откатил изменения в ветке master. После перемены в ветках, которую
сделал Сергей, рекомендуется использовать именно ветку master.

чт, 14 мар. 2019 г. в 13:07, Andrey <notifications@github.com>:

**Comment by aol-nnov** [comment]
@vatilin здравствуйте!

Сейчас перешел на мастер, да, шаманством заниматься не пришлось - таски не лочатся, но моё утверждение про маки осталось в силе.

могу пинговать борду и использоват tcp сервер только если установить мак таким, чтобы он при любом порядке байтов выглядел одинаково.

сейчас прописал `ee:ee:ee:ee:ee:ee` - работает. Осталось вот это выяснить...

**Comment by aol-nnov** [comment]
Дмитрий, еще вот какая мистика началась..

у нас две платы в разных отделах в разных городах (так что, я не смогу достоверно сказать, что там у соседей), но после сборки с a308284c5a8189ae9957670f63783a053109f9fd у них hard fault, а у меня всё хорошо (не считая истории с мак-адресом)...

**Comment by aol-nnov** [comment]
примеры, похоже, с проверкой знаний матчасти ;)

берем [вот это](https://github.com/sergev/uos-embedded/blob/master/examples/arm-1986be1/test-ve1eth.c#L207) и [вот это (из википедии)](https://en.wikipedia.org/wiki/MAC_address#/media/File:MAC-48_Address.svg), совмещаем. Изображаем на лице работу мысли, потом, вскинув палец вверх, восклицаем "ага!", меняем `0x01` на что-то не  содержащее 1 в младшем бите, чтобы не получился "multicast mac". и всё работает :)

зато попутно покопался в кишках uos-а. хитрый план у вас, однако! :-D

**Comment by aol-nnov** [comment]
@vatilin еще на мастере не хватает фикса для eth_debug - залочки мьютекса нет, а разлочка есть. в результате - паника :) (а на stable-eth-devel) он есть

**Comment by vatilin** [comment]
Обновил в ветке master драйвер Ethernet для Миландра. Коллега проверил на
их отладочной плате - работает.

чт, 14 мар. 2019 г. в 16:25, Andrey <notifications@github.com>:

**Comment by aol-nnov** [comment]
Дмитрий, позвольте уточнить, отладочная плата была подключена в непосредственно в компьютер или через коммутатор?

Наши эксперименты показывают, что мак-адреса, начинающиеся на 0x1-0x4 (как раз, [как в примере](
https://github.com/sergev/uos-embedded/blob/master/examples/arm-1986be1/test-ve1eth.c#L207)) "не работают". Причину я описал выше со ссылкой на стандарт. Есть подозрение, что коммутаторы фильтруют такие маки и не пропускают их. Я не спец в  802.1D, не могу заявлять точно по этому поводу.. 

Пришли к такому заключению опытным путём.

**Comment by vatilin** [comment]
До этого проверяли напрямую. Попробовали через относительно старый
коммутатор (Switch) от D-Link, тоже работает. Может быть, у вас какой-то
умный роутер?

пт, 15 мар. 2019 г. в 16:02, Andrey <notifications@github.com>:

**Comment by aol-nnov** [comment]
да, не сказать, чтобы умный.. но эффект воспроизводится в двух не связанных подразделениях. оборудование тоже разное. в общем, мистика какая-то осталась.

у меня D-Link DES-1210-28/ME с дефолтными настройками, например..
